### PR TITLE
[CSM-275] Add tx fee endpoint

### PR DESCRIPTION
**Pull request by pva701**
* Introduce endpoint for getting fee for transaction (with according to current fee policy)

### CODE REVIEW

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 1]
```diff
@@ -515,6 +535,34 @@ getMoneySourceWallet (AddressMoneySource addrId) = cwamWId addrId
...
+            TxFeePolicyUnknown w _               -> throwM $ unknownFeePolicy w
+            TxFeePolicyTxSizeLinear linearPolicy -> do
+                (srcAddrMetas, outs, _) <- prepareTxInfo passphrase moneySource dstDistr
+                sks <- forM srcAddrMetas $ getSKByAccAddr passphrase
+                srcAddrs <- forM srcAddrMetas $ decodeCIdOrFail . cwamId
+                txAux <- withSafeSigners passphrase sks $ \ss -> do
+                    let hdwSigner = NE.zip ss srcAddrs
+                    let addrs = map snd $ toList hdwSigner
+                    utxo <- getOwnUtxos addrs
+                    runExceptT $ ExceptT $ pure $ createMTx utxo hdwSigner outs
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: `runExceptT $ ExceptT $ ...`?

Comment (pva701) [inline src/Pos/Wallet/Web/Server/Methods.hs]: lol, mda uj

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 2]
```diff
@@ -515,6 +535,34 @@ getMoneySourceWallet (AddressMoneySource addrId) = cwamWId addrId
...
+        Nothing -> pure $ mkCoin 0
+        Just feePolicy -> case feePolicy of
+            TxFeePolicyUnknown w _               -> throwM $ unknownFeePolicy w
+            TxFeePolicyTxSizeLinear linearPolicy -> do
+                (srcAddrMetas, outs, _) <- prepareTxInfo passphrase moneySource dstDistr
+                sks <- forM srcAddrMetas $ getSKByAccAddr passphrase
+                srcAddrs <- forM srcAddrMetas $ decodeCIdOrFail . cwamId
+                txAux <- withSafeSigners passphrase sks $ \ss -> do
+                    let hdwSigner = NE.zip ss srcAddrs
+                    let addrs = map snd $ toList hdwSigner
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: I guess these 2 lines could be collapsed

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 3]
```diff
@@ -621,6 +692,8 @@ sendMoney sendActions passphrase moneySource dstDistr = do
                 "\n" %build)
                (toList entries)
                remains
+    listF separator formatter =
+        F.later $ fold . intersperse separator . fmap (F.bprint formatter)
```
Comment (Martoon-00) [inline src/Pos/Wallet/Web/Server/Methods.hs]: If you don't want to extract it at least place a todo

File: src/Pos/Wallet/Web/Server/Methods.hs (hunk 2 above)
Comment (pva701) [inline src/Pos/Wallet/Web/Server/Methods.hs]: no,`hdsSigner` is used below

**Comment by pva701** [comment]
@Martoon-00 I have force pushed

File: src/Pos/Wallet/Web/Server/Methods.hs [hunk 4]
```diff
@@ -44,6 +44,7 @@ import           Pos.ReportServer.Report          (ReportType (RInfo))
 import           Serokell.AcidState.ExtendedState (ExtendedState)
 import           Serokell.Util                    (threadDelay)
 import qualified Serokell.Util.Base64             as B64
+import           Pos.DB.Class                     (gsAdoptedBVData)
```
Comment (gromakovsky) [inline src/Pos/Wallet/Web/Server/Methods.hs]: 😒

Comment (pva701) [inline src/Pos/Wallet/Web/Server/Methods.hs]: sorry))))))))0000000
will fix within CSM-321 (working on it now)

Comment (gromakovsky) [inline src/Pos/Wallet/Web/Server/Methods.hs]: http://i.imgur.com/kz54HbA.png

Comment (pva701) [inline src/Pos/Wallet/Web/Server/Methods.hs]: this is a lie

**Comment by akegalj** [comment]
@pva701 we forgot to bind these endpoints to client-api (ie, purscript wrapper) that UI is using atm.

**Comment by pva701** [comment]
Oh, how to do it?
@Martoon-00 @flyingleafe can you assist?

**Comment by akegalj** [comment]
@pva701 don't lose time on this - I will wire it :)

File: src/Pos/Wallet/Web/Api.hs [hunk 5]
```diff
@@ -241,6 +241,15 @@ type NewPayment =
     :> Capture "amount" Coin
     :> WRes Post CTx
 
+type TxFee =
+       "txs"
+    :> "fee"
+    :> DCQueryParam "passphrase" CPassPhrase
```
Comment (akegalj) [inline src/Pos/Wallet/Web/Api.hs]: @pva701 do we really need to have password for calculating tx fee ?

Comment (pva701) [inline src/Pos/Wallet/Web/Api.hs]: @akegalj password is needed for creating of tx, I just used existed functions.
Fee depends on size of created tx.
@Martoon-00 we can't create a tx without a password, can we?

Comment (georgeee) [inline src/Pos/Wallet/Web/Api.hs]: @pva701 we sure don't need it. We can create tx right away, we can not sign it w/o password. But signature length is always same, so we just need to add it to size of tx itself.

BTW, this is interesting point. I believe tx fee should use `s = tx.size + witness.size + signature.size`

**Pull request merged**